### DISCUSSION TITLE
How best to query all the files in a project/solution?

### DISCUSSION THREAD

**Discussion by mrlacey** [VsixCommunity/Community.VisualStudio.Toolkit#9] [Q&A]
I have a few extensions that need to know about every file in the solution.
The SDK options for querying all the files in a solution are _really_ slow, so I query the file system directly. 
This is much faster (is acceptable) but has three potential issues:
- It can't handle files in the solution that are outside the root directory of the solution.
- It can't account for files that are in the filesystem but not part of the solution.
- It can't handle code spaces, live share, and other remote file systems that are coming with VS17.

Is there a better way?
How do other people handle this scenario?
Could the toolkit provide something to help here?

**Comment by reduckted** [comment]
> The SDK options for querying all the files in a solution are _really_ slow

What technique are you currently using?

The `IVSHierarchy` model is all COM-style APIs and is pretty horrible to use. Perhaps some wrapper classes for that would help?

**Reply by mrlacey** [reply]
I don't have the original code anymore, but was getting the Solution object, then for each Project (in `Solution.Projects`) I'd iterate through the `ProjectItems` in the project.
Having to allow for the different project types, recursion, and solution folder variants would be good in a helper wrapper class but even when I did this I still found it slow for large projects/solutions. I think a key part of the performance issue was that accessing some of the underlying types needs to be done on the UI/main thread and so the delay was significant. 
In contrast, it's easy to get the directory the solution file is in and then search for files with a specific extension, and then use a FileSystemWatcher to pick up changes.  

As a separate issue/feature a wrapper for accessing _all_ projects in a solution could be a useful addition here (I have other extensions where I do this) but my bigger interest is iterating through all the files in a project/solution.

**Reply by StevenRasmussen** [reply]
I don't have much to add here other than, "yeah... this is really painful".  I'm essentially doing as @mrlacey laid out by iterating over all the projects and their items in a solution.  I haven't tried using the `IVSHierarchy` approach.  Do you know if it's more performant than using the `DTE` objects?  If so, I could look at writing some wrappers/extension methods and contributing them to the project.
